packages/api-server/src/db/query.ts

```diff
@@ -299,22 +302,37 @@ export class Query {
   private async queryLogsByBlockRange(
     fromBlock: HexNumber,
     toBlock: HexNumber,
-    address?: HexString | HexString[],
+    queryAddresses: HexString[],
+    queryTopics: FilterTopic[],
     lastPollId?: bigint,
     offset?: number
   ): Promise<Log[]> {
     const { MAX_QUERY_NUMBER } = getDatabaseRateLimitingConfiguration();
     const queryLastPollId = lastPollId || -1;
     const queryOffset = offset || 0;
-    let logs: DBLog[] = await this.knex<DBLog>("logs")
-      .modify(buildQueryLogAddress, address)
+
+    // NOTE: In this SQL, there is no `ORDER BY id` as combining `ORDER BY` and `LIMIT` consumes too much time when the
+    // results are large. Instead, logs are sorted outside the database:
+    // - When the number of query results exceeds $MAX_QUERY_NUMBER, an error is returned.
+    // - When the queried results are less than or equal to $MAX_QUERY_NUMBER, sorting takes only a short time
+    //
+    // NOTE: Using CTE to SELECT "logs" then JOIN "transactions" is more efficient than directly SELECT JOIN
+    let selectLogs = this.knex<DBLog>("logs")
+      .modify(buildQueryLogAddress, queryAddresses)
+      .modify(buildQueryLogTopics, queryTopics)
       .where("block_number", ">=", fromBlock)
       .where("block_number", "<=", toBlock)
       .where("id", ">", queryLastPollId.toString(10))
-      .orderBy("id", "asc")
       .offset(queryOffset)
```

Comment:
Could `offset` be removed? Using `offset` without `order by` is problematic.

Reply:
Yes, I agree!
Due to the impact of removing `offset` on other interfaces, I will create some new commits to refactor these interfaces, such as `eth_getFilterChanges`.